### DISCUSSION TITLE
Weird sporadic KeyError in path_registry

### DISCUSSION THREAD

**Discussion by kasium** [sqlalchemy/sqlalchemy#9745] [Usage Questions]
Hi all,
first, I'm very sorry for the limited information I can provide. I'm aware that this makes the support more complicated, but I tried as much as possible to grab all Information, but nothing helped. So, here we go;

In my system, I can see this error from time to time. The system uses flask, flask-sqlalchemy and SQLAlchemy (1.3). However, as you can see in the stack it relates to the internals of SQLAlchemy. The system gets thousands of requests each day but the error only occurred 27 times in the last 30 days. It could be related to a very specific thread-pattern or the python garbage collection. 

The code at this point is
```py
    def get(self, attributes, key, value=None):
        key = (key, self.natural_path)
        if key in attributes:
            return attributes[key]
        else:
            return value
```
Since it fails in the return line and the `in` checked passed, I either suspect a weird modification of `attributes` or a garbage collection issue.

The full stack trace:
```
KeyError: ('memoized_setups', (<Mapper at 0x7f5208fb2190; MyTable>,))
  File "flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "dummy.py", line 1418, in dummy
    return jsonify(obj.json(attributes=attributes))
  File "dummy.py", line 230, in json
    for status_obj in self.status
  File "sqlalchemy/orm/attributes.py", line 294, in __get__
    return self.impl.get(instance_state(instance), dict_)
  File "sqlalchemy/orm/attributes.py", line 730, in get
    value = self.callable_(state, passive)
  File "sqlalchemy/orm/strategies.py", line 759, in _load_for_state
    return self._emit_lazyload(
  File "<string>", line 1, in <lambda>
  File "sqlalchemy/orm/strategies.py", line 903, in _emit_lazyload
    .all()
  File "sqlalchemy/ext/baked.py", line 544, in all
    return list(self)
  File "sqlalchemy/orm/loading.py", line 99, in instances
    with util.safe_reraise():
  File "sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "sqlalchemy/orm/loading.py", line 58, in instances
    *[
  File "sqlalchemy/orm/loading.py", line 59, in <listcomp>
    query_entity.row_processor(query, context, cursor)
  File "sqlalchemy/orm/query.py", line 4422, in row_processor
    _instance = loading._instance_processor(
  File "sqlalchemy/orm/loading.py", line 386, in _instance_processor
    quick_populators = path.get(
  File "sqlalchemy/orm/path_registry.py", line 93, in get
    return attributes[key]
```

I have no way to reproduce it. Therefore just my ask, if you have any idea and if e.g. the KeyError could be caught? The issue looks somehow like https://github.com/sqlalchemy/sqlalchemy/issues/4030

**Comment by CaselIT** [comment]
Hi,

Interestingly the same code is still present:
https://github.com/sqlalchemy/sqlalchemy/blob/db9a2caa43f0e8539bd1b3d8a2522f8018903605/lib/sqlalchemy/orm/path_registry.py#L167-L174

We could probably replace it with `return attributes.get((key, self.natural_path), value)` that should avoid this particular race condition

Still, it's strange that the attributes dict would mutate like this. It's unlikely it's gc  related, since it's a plain dict (at least in v2), that's keeps strong refs of key/values.

**Reply by kasium** [reply]
Thanks a lot! Sounds a like a good way forward

**Reply by kasium** [reply]
@CaselIT should we transfer this to an issue?

**Reply by CaselIT** [reply]
@zzzeek do you have suggestions here?

**Reply by zzzeek** [reply]
yes I have a super awesome suggestion, this is the old ORM lazy loader baked queries use.  turn it off in the relationship that causes the problem:

```py
class MyClass(Base):
    # ...

   some_collection = relationship(Thing, ... bake_queries=False)
```

seek to upgrade to SQLAlchemy 1.4 which will also not have this problem since baked queries are no longer used.

**Reply by kasium** [reply]
Thanks, upgrade to 1.4 and then 2.0 is already planned!

**Reply by zzzeek** [reply]
> Hi,
> 
> Interestingly the same code is still present:
> 
> https://github.com/sqlalchemy/sqlalchemy/blob/db9a2caa43f0e8539bd1b3d8a2522f8018903605/lib/sqlalchemy/orm/path_registry.py#L167-L174
> 
> We could probably replace it with `return attributes.get((key, self.natural_path), value)` that should avoid this particular race condition

this code is not expected to be threadsafe for "attributes".    attributes is fully local to the compilation context now.   

in 1.3, using bakedqueries, the .attributes is sitting on the Query object that's being used in a concurrent fashion.    While this Query object is in the LRU, and can get pulled out of the LRU at any time, GC doesn't explain this because that "attributes" dictionary is strongly referenced here.  Something would have to be removing from it.   There is exactly one place where that happens in the 1.3 codebase and it is specific to 1. baked queries and 2. subquery loaders.   so the issue is likely that combination of loaders.

in 1.4/2.0, there are many levels of "none of this happens" - "baked queries" are just a facade on top of normal cached queries now, there is no "del" from .attributes, .attributes` is local to a SQLCompiler object, and subqueryloading is legacy now.

@kasium another thing to look into is not using "subqueryload" in favor of "selectinload", which might also remove this problem.
